### PR TITLE
bpo-37128: Add math.perm().

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue37128

### CODE REVIEW

**Comment by mdickinson** [comment]
@serhiy-storchaka Please could you open a new b.p.o. issue for `math.perm`? I think we do need to give people an opportunity to discuss before it goes in, and since issue 35431 ended up focusing on `math.comb`, people interested in `math.perm` might not be following it.

File: Doc/library/math.rst [hunk 1]
```diff
@@ -192,6 +207,18 @@ Number-theoretic and representation functions
    of *x* and are floats.
 
 
+.. function:: perm(n, k)
+
+   Return the number of ways to choose *k* items from *n* items without repetition.
```
Comment (rhettinger) [inline Doc/library/math.rst]: The "with order" part should be mentioned.  Perhaps ``s/the number/the distinct number/``.

FWIW, here is the wording from the *itertools* docs:


```
permutations() | p[, r] | r-length tuples, all possible orderings, no repeated elements

combinations() | p, r | r-length tuples, in sorted order, no repeated elements
```

File: Modules/mathmodule.c [hunk 2]
```diff
@@ -2998,27 +2998,8 @@ math_prod_impl(PyObject *module, PyObject *iterable, PyObject *start)
...
-
-Raises TypeError if the arguments are not integers.
-Raises ValueError if the arguments are negative or if k > n.
-
-[clinic start generated code]*/
-
 static PyObject *
-math_comb_impl(PyObject *module, PyObject *n, PyObject *k)
-/*[clinic end generated code: output=bd2cec8d854f3493 input=2f336ac9ec8242f9]*/
+perm_comb(PyObject *n, PyObject *k, int comb)
```
Comment (rhettinger) [inline Modules/mathmodule.c]: ISTM that combining the code for *comb()* and *perm()* doesn't save us much.  I would rather they be decoupled at outset so as to not preclude possible optimizations (there are a number of ways to compute combinations that do early cancellation to keep the intermediate products from exploding in size).

**Review by rhettinger — APPROVED** [review]

File: Doc/library/math.rst [hunk 3]
```diff
@@ -207,6 +207,19 @@ Number-theoretic and representation functions
    of *x* and are floats.
 
 
+.. function:: perm(n, k)
+
+   Return the distinct number of ways to choose *k* items from *n* items
```
Comment (mdickinson) [inline Doc/library/math.rst]: "distinct number of ways" doesn't quite make sense. "number of distinct ways"? Or just lose the "distinct" word altogether.

File: Modules/mathmodule.c [hunk 4]
```diff
@@ -3028,11 +3142,24 @@ math_comb_impl(PyObject *module, PyObject *n, PyObject *k)
     if (n == NULL) {
         return NULL;
     }
+    if (!PyLong_CheckExact(n)) {
```
Comment (mdickinson) [inline Modules/mathmodule.c:3145]: I'm looking forward to the day when we can remove this kind of block as unnecessary (perhaps in 3.9). I thought I'd persuaded myself that everything does eventually get converted to plain `int`s, but now that I look again it's not clear, and I agree that making this check up front is a good idea.

**Review by mdickinson — APPROVED** [review]
LGTM; one request for a docstring tweak.

**Pull request merged**